### PR TITLE
Revise the socket implementation of c10d

### LINKED ISSUE #52040: torch.distributed.TCPStore doesn't work with dual IPv4/IPv6 network interface
## To Reproduce

1. Setup 2 machines with dual IPv4/IPv6 interfaces
2. on first machine runthe following script
```
import torch.distributed as dist
from datetime import timedelta
import socket

port = 12345
ip = "127.0.0.1"
ip = "::"
try:
    server_store = dist.TCPStore(ip, port, 2, True, timedelta(seconds=30))
    print("initialized")
    server_store.set("first_key", "first_value")
    print("stored")
except RuntimeError as e:
    print(e)

```
3. on second machine run the following script
```
import torch.distributed as dist
from datetime import timedelta
import socket

port = 12345
ip = "172.0.0.1"
ip = "xxxx:yyyy:::"
client_store = dist.TCPStore(ip, port, 2, False, timedelta(seconds=30))
print("initialized")
print(client_store.get("first_key"))
```
4. Observe that this code works for IPv4 (127.0.0.1 on first, ipv4 of first machine on second) but doesn't work when IPv6 address is used on master node. Second/subordinate node uses IPv6 correctly.
5. Proposed fix:
    In https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/Utils.cpp#L90
    in line ::getaddrinfo nullptr is passed as a first argument, instead masteraddr should probably be passed down from outer callstack. We verified that if we replace nullptr with "::" the code works.

## Expected behavior

TCPStore should work with both IPv4 and IPv6 addresses the same way.

## Environment

Network interface with ipv4 and ipv6 address.

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7.3
 - CUDA/cuDNN version: 11.2, 8.0
 - GPU models and configuration: V100
 - Any other relevant information:

### LINKED ISSUE #55571: Update TCPStore.wait() error message to be more friendly
```python
import torch.distributed as dist
server_store = dist.TCPStore("127.0.0.1", 1234, is_master=True, timeout=timedelta(seconds=5))
server_store.wait(["invalid_key"])
```

output
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Socket Timeout
```

### LINKED ISSUE #63802: [torch.distributed.launch|run] Hangs on SIGINT when using a TCPStore backed rdzv_backend
## 🐛 Bug

`torch.distributed.launch` and `torch.distributed.run` will not process SIGINT (CTRL+C from CLI) and hang indefinitely without returning control back to the shell when trying to CTRL+C out of the launcher while it is waiting for other nodes to show up (aka during rendezvous)

## To Reproduce

1. (install torch 1.9.0+, issue exists in torch-1.10 nightly as well)
2. Force an incomplete rank with rdzv (`--nnodes 2` but just run the following command once)
    ```shell
     $ LOGLEVEL=INFO python -u -m torch.distributed.run \
                --nproc_per_node 4 \
                --nnodes 2 \
                --master_addr $(hostname) \
                --master_port 6000 \
                ~/test_script.py
    ```
3. (note)  content of `~/test_script.py` does not matter (`torch.distributed.run` gets stuck before the script content is even relevant). But here's what I tested with anyways
    ```py
    import torch.distributed as dist
    from torch.distributed.elastic.multiprocessing.errors import record
    import argparse
    import torch

    @record
    def main():
        import os
        rank = os.environ["RANK"]
        dist.init_process_group("nccl")

    if __name__ == "__main__":
        main()
    ```
4.  #2's command will hang so try CTRL+C'ing it
5. (actual) hangs indefinitely (see screenshot below)

image

6. Sending it SIGINT (`kill -15 $pid`) or SIGKILL (`kill -9 $pid`) from a different terminal works but inconvenient since I lose control over the main terminal

## Expected behavior

CTRL+C should terminate `torch.distributed.run` and return the control back to the main terminal

## Environment

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

PyTorch version: 1.10.0.dev20210823+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.7 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.23

Python version: 3.8.5 (default, Mar 24 2021, 22:32:23)  [GCC 5.4.0 20160609] (64-bit runtime)
Python platform: Linux-4.4.0-1102-aws-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB

Nvidia driver version: 440.33.01
cuDNN version: /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.20.3
[pip3] pytorch-lightning==1.3.5
[pip3] torch==1.10.0.dev20210823+cu102
[pip3] torchaudio==0.9.0
[pip3] torchelastic==0.2.3.dev0
[pip3] torchmetrics==0.3.2
[pip3] torchserve==0.4.0
[pip3] torchvision==0.10.0+cu111
[conda] Could not collect

## Additional context

SIGINT seems to be ignored during the creation of TCPStore (which we use for "static" and "c10d" rdzv backends) since the main thread/process is in a tcp accept() call which seems to not be "interruptable". https://stackoverflow.com/questions/44259468/how-to-interrupt-accept-in-a-tcp-ip-server

Using other (etcd) rendezvous backend does not have this issue.

### PR DESCRIPTION
**Pull request by cbalioglu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68226

**Note that this PR is unusually big due to the urgency of the changes. Please reach out to me in case you wish to have a "pair" review.**

This PR introduces a major refactoring of the socket implementation of the C10d library. A big portion of the logic is now contained in the `Socket` class and a follow-up PR will further consolidate the remaining parts. As of today the changes in this PR offer:

 - significantly better error handling and much more verbose logging (see the example output below)
 - explicit support for IPv6 and dual-stack sockets
 - correct handling of signal interrupts
 - better Windows support

A follow-up PR will consolidate `send`/`recv` logic into `Socket` and fully migrate to non-blocking sockets.

## Example Output

```
[I logging.h:21] The client socket will attempt to connect to an IPv6 address on (127.0.0.1, 29501).
[I logging.h:21] The client socket is attempting to connect to [localhost]:29501.
[W logging.h:28] The server socket on [localhost]:29501 is not yet listening (Error: 111 - Connection refused), retrying...
[I logging.h:21] The server socket will attempt to listen on an IPv6 address.
[I logging.h:21] The server socket is attempting to listen on [::]:29501.
[I logging.h:21] The server socket has started to listen on [::]:29501.
[I logging.h:21] The client socket will attempt to connect to an IPv6 address on (127.0.0.1, 29501).
[I logging.h:21] The client socket is attempting to connect to [localhost]:29501.
[I logging.h:21] The client socket has connected to [localhost]:29501 on [localhost]:42650.
[I logging.h:21] The server socket on [::]:29501 has accepted a connection from [localhost]:42650.
[I logging.h:21] The client socket has connected to [localhost]:29501 on [localhost]:42722.
[I logging.h:21] The server socket on [::]:29501 has accepted a connection from [localhost]:42722.
[I logging.h:21] The client socket will attempt to connect to an IPv6 address on (127.0.0.1, 29501).
[I logging.h:21] The client socket is attempting to connect to [localhost]:29501.
[I logging.h:21] The client socket has connected to [localhost]:29501 on [localhost]:42724.
[I logging.h:21] The server socket on [::]:29501 has accepted a connection from [localhost]:42724.
[I logging.h:21] The client socket will attempt to connect to an IPv6 address on (127.0.0.1, 29501).
[I logging.h:21] The client socket is attempting to connect to [localhost]:29501.
[I logging.h:21] The client socket has connected to [localhost]:29501 on [localhost]:42726.
[I logging.h:21] The server socket on [::]:29501 has accepted a connection from [localhost]:42726.
```

Differential Revision: [D32372333](https://our.internmc.facebook.com/intern/diff/D32372333/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32372333/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

**Comment by aivanou** [comment]
This looks great! I was wondering maybe it makes sense to create a completely new TCPStore? And modify this module https://github.com/pytorch/pytorch/blob/master/torch/distributed/rendezvous.py#L134

to add a new parameter to the rendezvous, e.g. `use_new_store`, that would control whether we should use new TCPStore or not? In this way, users will be able to fall back to the old implementation if bc changes will be detected.

File: test/distributed/elastic/utils/distributed_test.py
Comment (mrshenli) [inline test/distributed/elastic/utils/distributed_test.py:145]: Are these considered BC-breaking changes? It looks OK to me. We just need to add a [BC-Breaking] prefix to the title so that we remember to mention it in the release commit triage docs.

File: torch/csrc/distributed/c10d/socket.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/socket.cpp:639]: Log the connect_timeout here as well?

File: torch/csrc/distributed/c10d/socket.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/socket.cpp:604]: it looks like we keep retrying until timeout, is it worth it to also log how many times we've now retried?

File: test/distributed/test_store.py
Comment (mrshenli) [inline test/distributed/test_store.py:160]: Are we going to print more precise error messages else where regarding what the failure is? Besides, does this mean we expect a different error here than "Address already in use"?

**Comment by cbalioglu** [comment]
> This looks great! I was wondering maybe it makes sense to create a completely new TCPStore? And modify this module https://github.com/pytorch/pytorch/blob/master/torch/distributed/rendezvous.py#L134
> 
> to add a new parameter to the rendezvous, e.g. `use_new_store`, that would control whether we should use new TCPStore or not? In this way, users will be able to fall back to the old implementation if bc changes will be detected.

Thanks @aivanou! I totally agree that we should have a new implementation of `TCPStore`. The changes included in this PR are at the socket level though and should not affect the behavior of the existing `TCPStore` (except better error messages). Once we bring the C10d socket implementation into good shape, the next step would be to revise our stores.

File: test/distributed/test_store.py
Comment (cbalioglu) [reply test/distributed/test_store.py:160]: The regex just tests the start of the error string. A full error string that the user will see, would look like:
```
The server socket has failed to listen on any local network address. The server socket has failed to bind to [::]:29501 (generic error: 98 - Address already in use). The server socket has failed to bind to 0.0.0.0:29501 (generic error: 98 - Address already in use).
```

File: test/distributed/elastic/utils/distributed_test.py
Comment (cbalioglu) [reply test/distributed/elastic/utils/distributed_test.py:145]: `create_c10d_store` is an internal function used by the TorchElastic agent. It is not a public API, so we should be safe here.

File: torch/distributed/elastic/utils/distributed.py
Comment (H-Huang) [inline torch/distributed/elastic/utils/distributed.py:19]: Looks like `test_tcp_store_timeout_set` reads the `connect() timed out` message to determine whether or not to retry on failure. May also need to update this test to prevent it from being flaky.

File: torch/csrc/distributed/c10d/socket.cpp
Comment (cbalioglu) [reply torch/csrc/distributed/c10d/socket.cpp:604]: From user's perspective we try it only once. These three error conditions only happen if the server is not reachable. Once it becomes reachable (we internally retry every second), the `connect()` syscall will take the rest of the function and use `poll()` with the remaining timeout value. So there is really not a clear way to communicate what we mean by an "internal retry". If the server is reachable, but `connect()` hangs, we may not even have a retry.

File: torch/csrc/distributed/c10d/socket.cpp
Comment (mrshenli) [inline torch/csrc/distributed/c10d/socket.cpp:115]: I don't have objections to the current solution, just wanna mention the past discussion on this. 

https://github.com/pytorch/pytorch/pull/47749 introduced a few windows-only files to minimize the number of `#ifdef` in the code. Looks like this PR is going towards the opposite direction. It's totally your call, especially given that you own this part of our code base. :)

File: torch/csrc/Exceptions.h
Comment (mrshenli) [inline torch/csrc/Exceptions.h:82]: curious about how did we determine the order of the catch clauses here?

File: torch/csrc/distributed/c10d/socket.cpp
Comment (mrshenli) [inline torch/csrc/distributed/c10d/socket.cpp:194]: curious, how did we decide to use this error message for `invalid_argument` error instead of saying "invalid argument"? Is this the convention for socket error handling?

File: torch/csrc/distributed/c10d/socket.cpp
Comment (mrshenli) [inline torch/csrc/distributed/c10d/socket.cpp:245]: if this won't happen as an immediate followup PR, shall we create an issue to track this?

File: torch/csrc/distributed/c10d/socket.cpp
Comment (mrshenli) [inline torch/csrc/distributed/c10d/socket.cpp:381]: nit: `{}network` -> `{} network` ?

**Review by aivanou — COMMENTED** [review]
Should we remove this comment if it is not applicable anymore? Or should be add back `closeSocket` ?

File: torch/csrc/distributed/c10d/socket.cpp
Comment (cbalioglu) [reply torch/csrc/distributed/c10d/socket.cpp:194]: Below an excerpt from `man 2 accept`:

```
RETURN VALUE
  ...
  EINVAL Socket is not listening for connections, or addrlen is invalid (e.g., is negative).
```

Since we know for sure that `addrlen` is valid (see line 183), our error message correctly conveys the information.

**Review by mrshenli — APPROVED** [review]
Stamp to unblock. Please wait for tests.

File: torch/csrc/distributed/c10d/socket.cpp
Comment (cbalioglu) [reply torch/csrc/distributed/c10d/socket.cpp:245]: Hope to continue working on this project and finish it before the end of the year. I think it makes sense to have a task anyways though for keeping record things.

File: torch/csrc/distributed/c10d/socket.cpp
Comment (cbalioglu) [reply torch/csrc/distributed/c10d/socket.cpp:381]: It is in fact `{}network` and the possible strings are `IPv4 `, `IPv6 `, and ``. Because we need to have an empty string, a space between `{}` and `network` would cause double space.

File: torch/csrc/Exceptions.h
Comment (cbalioglu) [reply torch/csrc/Exceptions.h:82]: I believe we can move `TimoutException` below `c10::Error`, since all other clauses above it are subclasses of `c10::Error`. We can have `TimeoutException` and `system_error` at the end of the list.

File: torch/csrc/Exceptions.h
Comment (suo) [reply torch/csrc/Exceptions.h:12]: what happens when this code is compiled with USE_DISTRIBUTED=0?

File: torch/csrc/Exceptions.h
Comment (suo) [inline torch/csrc/Exceptions.h:84]: Can we make a `c10::TimeoutError` that corresponds to `PyExc_TimeoutError` and use that in c10d, instead of defining the error type in c10d and including it here?

**Comment by eric-haibin-lin** [comment]
Great work! @cbalioglu could you share how to turn on C10D_INFO logs? I see there're gflags for `caffe2_log_level`, but how do I turn them on with a nightly build pytorch running model training in python?

**Comment by cbalioglu** [comment]
Hey @eric-haibin-lin! Good to hear from an old colleague :) Unfortunately as of today if you want to change the log level, you have to build PyTorch with GLOG and use the `GLOG_minloglevel` environment variable. We plan to lift this limitation since this requires building PyTorch from scratch which is a pretty high bar for many of our users. You can follow the relevant conversation in #70667.

**Comment by cbalioglu** [comment]
@eric-haibin-lin Check out #73357. It is part of v1.11 meaning you can change the log level via the `TORCH_CPP_LOG_LEVEL` environment variable.

**Pull request closed without merging**
